kernel/ipi.c

```diff
@@ -13,15 +13,47 @@ extern void z_trace_sched_ipi(void);
 #endif
 
 
-void flag_ipi(void)
+void flag_ipi(uint32_t ipi_mask)
 {
 #if defined(CONFIG_SCHED_IPI_SUPPORTED)
 	if (arch_num_cpus() > 1) {
-		_kernel.pending_ipi = true;
+		atomic_or(&_kernel.pending_ipi, (atomic_val_t)ipi_mask);
 	}
 #endif /* CONFIG_SCHED_IPI_SUPPORTED */
 }
 
+/* Create a bitmask of CPUs that need an IPI. Note: sched_spinlock is held. */
+atomic_val_t ipi_mask_create(struct k_thread *thread)
+{
+#if defined(CONFIG_IPI_OPTIMIZE)
+	uint32_t  ipi_mask = 0;
+	uint32_t  num_cpus = (uint32_t)arch_num_cpus();
+	uint32_t  id = _current_cpu->id;
+	struct k_thread *cpu_thread;
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
+		if (id == i) {
+			continue;
+		}
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
```

Comment:
FWIW, this logic still isn't quite right, though it definitely seems to be wrong in the right/conservative direction: it will flag IPIs in the case where the tested thread is higher priority than the foreign thread always, even if the thread can't run on the other CPU because it's running a cooperative thread and/or disallowed via cpu_mask.

Reply:
If I am following you correctly, at this point in time, that is intentional. The overall intent was not to provide a perfect prediction of what CPUs need an IPI, but to determine in as cheap a manner as practical only if it _might_ need one. Receiving an unnecessary IPI does impact performance (which we are trying to improve), but ought not to impact correctness of scheduling and execution.

Reply:
I was originally thinking of refining this check at a later time, but it looks like it should be straight forward (I was originally worried that there might possibly be some places where we were not protecting things as expected). However, those fears appeared to be unfounded.

One implication is that I have to redo the test suite as CPUs running a cooperative thread will no longer receive an IPI due to a thread being made ready.